pkg/controller/certificates/internal/secretsmanager/secret.go

```diff
@@ -62,104 +70,70 @@ type SecretData struct {
-			},
-			Type: corev1.SecretTypeTLS,
-		}
-	}
-
-	// secret will be overwritten by 'existingSecret' if existingSecret is non-nil
-	if s.enableSecretOwnerReferences {
-		secret.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(crt, certificateGvk)}
-	}
-
-	secret = secret.DeepCopy()
-	err = s.setValues(crt, secret, data)
+	secret, err := s.getCertificateSecret(ctx, crt)
 	if err != nil {
 		return err
 	}
 
-	// If secret does not exist then create it
-	if !secretExists {
+	log := logf.FromContext(ctx).WithName("secrets_manager")
+	log = logf.WithResource(log, secret)
 
-		_, err = s.kubeClient.CoreV1().Secrets(secret.Namespace).Create(ctx, secret, metav1.CreateOptions{})
+	if err := s.setValues(crt, secret, data); err != nil {
 		return err
 	}
 
-	// Currently we are always updating. We should devise a way to not have to call an update if it is not necessary.
-	_, err = s.kubeClient.CoreV1().Secrets(secret.Namespace).Update(ctx, secret, metav1.UpdateOptions{})
-	return err
-}
+	// Build Secret apply configuration and options.
+	applyOpts := metav1.ApplyOptions{FieldManager: s.userAgent}
+	applyCnf := applycorev1.Secret(secret.Name, secret.Namespace).
+		WithAnnotations(secret.Annotations).WithLabels(secret.Labels).
+		WithData(secret.Data).WithType(secret.Type)
 
-func updateSecretWithAdditionalOutputFormats(crt *cmapi.Certificate, secret *corev1.Secret, data SecretData) error {
-	if crt.Spec.AdditionalOutputFormats == nil {
-		delete(secret.Data, cmapi.AdditionalOutputFormatDERKey)
-		delete(secret.Data, cmapi.AdditionalOutputFormatPEMKey)
-		return nil
+	// If Secret owner reference is enabled, set it on the Secret. This results
+	// in a no-op if the Secret already exists and has the owner reference set,
+	// and visa-versa.
+	if s.enableSecretOwnerReferences {
+		ref := *metav1.NewControllerRef(crt, certificateGvk)
+		applyCnf = applyCnf.WithOwnerReferences(&applymetav1.OwnerReferenceApplyConfiguration{
+			APIVersion: &ref.APIVersion, Kind: &ref.Kind,
+			Name: &ref.Name, UID: &ref.UID,
+			Controller: ref.Controller, BlockOwnerDeletion: ref.BlockOwnerDeletion,
+		})
 	}
 
-	additionalOutputFormatDER := false
-	additionalOutputFormatPEM := false
+	log.V(logf.DebugLevel).Info("applying secret")
 
-	for _, f := range crt.Spec.AdditionalOutputFormats {
-		switch f.Type {
-		case cmapi.AdditionalOutputFormatDER:
-			additionalOutputFormatDER = true
-		case cmapi.AdditionalOutputFormatCombinedPEM:
-			additionalOutputFormatPEM = true
-		default:
-			return fmt.Errorf("unknown additional output format %s", f.Type)
-		}
+	// Apply secret resource. Don't force apply first, so we can catch the error
+	// and log it.
+	_, err = s.secretClient.Secrets(secret.Namespace).Apply(ctx, applyCnf, applyOpts)
+	if apierrors.IsConflict(err) {
+		log.Error(err, "forcing apply due to field management conflict")
+		applyOpts.Force = true
+		_, err = s.secretClient.Secrets(secret.Namespace).Apply(ctx, applyCnf, applyOpts)
 	}
```

Comment:
Not sure why but it seems like the controller does three consecutive `PATCH` calls for this Apply operation:

![Screenshot from 2022-01-15 19-55-29](https://user-images.githubusercontent.com/2195781/149634409-4700db09-7c9a-4cae-bdf9-fd395abbbed2.png)

The cert-manager logs also seem to suggest three Apply operations:

```
I0115 19:47:24.073414   34135 secret_manager.go:82] cert-manager/certificates-issuing "msg"="mismatch between SecretTemplate and Secret, updating Secret annotations/labels" "key"="default/example2" "resource_kind"="Certificate" "resource_name"="example2" "resource_namespace"="default" "resource_version"="v1" "secret"="example2"
I0115 19:47:24.102769   34135 secret_manager.go:82] cert-manager/certificates-issuing "msg"="mismatch between SecretTemplate and Secret, updating Secret annotations/labels" "key"="default/example2" "resource_kind"="Certificate" "resource_name"="example2" "resource_namespace"="default" "resource_version"="v1" "secret"="example2"
I0115 19:47:24.113205   34135 secret_manager.go:82] cert-manager/certificates-issuing "msg"="mismatch between SecretTemplate and Secret, updating Secret annotations/labels" "key"="default/example2" "resource_kind"="Certificate" "resource_name"="example2" "resource_namespace"="default" "resource_version"="v1" "secret"="example2"
```

Is this expected?

Reply:
I don't expect this behaviour no. I'll reproduce this and see where this is coming from.

Reply:
Thanks for spot. The problem is here https://github.com/jetstack/cert-manager/pull/4638/files#diff-b58b0eda032fce40ccf63b138d72c40f2a5b1268b5dd675d9f99119957e2e9e8R175

We are doing a length check against the length of managed annotations against that of the SecretTemplate. This is wrong of course, since the `SecretsManager` will add the base annotations to the Secret managed by the issuing controller, so the length will never match. 

I'll add a function to return the set of base annotations that get added to Secrets and add that length to the if statement here.

Reply:
I wonder, why didn't the following unit test fail due to this?

```go
		"if secret does exist, update existing Secret and leave custom annotations, with owner disabled": {}
```

Is this tested somewhere else?

Reply:
Ah, you just added the tests I was thinking about, namely those three:

```go
"if managed fields matches template and base cert-manager annotations are present with no certificate data, should return true"
"if managed fields matches template and base cert-manager annotations are present with certificate data, should return true"
"if managed fields matches template and base cert-manager annotations are present with certificate data but certificate data is nil, should return false"
```

Thanks!